crates/energy/src/graph.rs

```diff
@@ -0,0 +1,259 @@
+use de_spawner::{DespawnEventsPlugin, DespawnedComponentsEvent};
+use parry3d::bounding_volume::Aabb;
+use parry3d::math::Point;
+use tinyvec::TinyVec;
+
+// The max distance (in meters) between two entities for them to be consider neighbors in the graph
+const MAX_DISTANCE: f32 = 10.0;
+
+pub(crate) struct GraphPlugin;
+
+impl Plugin for GraphPlugin {
+    fn build(&self, app: &mut App) {
+        app.add_plugins(DespawnEventsPlugin::<&NearbyUnits, NearbyUnits>::default())
+            .add_systems(OnEnter(GameState::Playing), setup)
+            .add_systems(OnExit(GameState::Playing), clean_up)
+            .add_systems(PostUpdate, spawn_graph_components)
+            .add_systems(
+                FixedUpdate,
+                (
+                    remove_old_nodes.before(GraphSystemSet::UpdateNearby),
+                    update_nearby_recv.in_set(GraphSystemSet::UpdateNearby),
+                    update_graph
+                        .in_set(GraphSystemSet::UpdateGraph)
+                        .after(GraphSystemSet::UpdateNearby),
+                )
+                    .run_if(in_state(GameState::Playing)),
+            );
+
+        #[cfg(feature = "energy_graph_debug_lines")]
+        app.add_plugins(DebugLinesPlugin::default()).add_systems(
+            PostUpdate,
+            (debug_lines).run_if(in_state(GameState::Playing)),
+        );
+    }
+}
+
+/// wrapped entity to allow for default values (se we can work with TinyVec)
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct NearbyEntity(Entity);
+
+impl Default for NearbyEntity {
+    fn default() -> Self {
+        Self(Entity::PLACEHOLDER)
+    }
+}
+
+impl From<NearbyEntity> for Entity {
+    fn from(val: NearbyEntity) -> Self {
+        val.0
+    }
+}
+
+impl From<Entity> for NearbyEntity {
+    fn from(entity: Entity) -> Self {
+        Self(entity)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, SystemSet)]
+enum GraphSystemSet {
+    UpdateNearby,
+    UpdateGraph,
+}
+
+/// The power grid resource is used to store the power grid graph.
+#[derive(Resource, Debug, Clone)]
+pub(crate) struct PowerGrid {
+    /// The power grid graph.
+    graph: GraphMap<Entity, f64, Undirected>,
+}
+
+impl Default for PowerGrid {
+    fn default() -> Self {
+        Self {
+            graph: GraphMap::new(),
+        }
+    }
+}
+
+/// The energy receiver component is used to mark an entity as an energy receiver.
+#[derive(Component, Debug, Clone, Copy)]
+pub struct EnergyReceiver;
+
+/// The energy producer component is used to mark an entity as an energy producer.
+#[derive(Component, Debug, Clone, Copy)]
+pub struct EnergyProducer;
+
+/// The nearby component is used to store the nearby entities of an entity.
+#[derive(Debug, Clone, Default)]
+pub struct Nearby {
+    receivers: TinyVec<[NearbyEntity; 32]>,
+    producers: TinyVec<[NearbyEntity; 32]>,
+}
+
+impl Nearby {
+    fn both(&self) -> TinyVec<[NearbyEntity; 64]> {
+        self.receivers
```

Comment:
I would rather return an iterator for performance reason (no need to construct a new `TinyVec` every time this is called). Alternatively, it can take a `&mut Vec` as an argument.

Note that I suggest to not separate producers & receivers (i.e. to have only single vec) in another comment. If we go that way, the method would not be needed at all.